api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,273 @@
+    HttpBody response_body = 5;
+
+    // The HTTP trailers for the request path. These are only sent if
+    // the server requested them. If "response_required" is set, the server
+    // must send back a TrailerResponse message or close the stream.
+    HttpTrailers request_trailers = 6;
+
+    // The HTTP trailers for the response path. These are only sent if
+    // the server requested them. If "response_required" is set, the server
+    // must send back a TrailerResponse message or close the stream.
+    HttpTrailers response_trailers = 7;
+  }
+}
+
+// For every ProcessingRequest received by the server with the "response_required" field
+// set, the server must send back exactly one ProcessingResponse message.
+// In addition, the server may send back a message with the "async_mode_change"
+// message at any time.
+// [#next-free-field: 11]
+message ProcessingResponse {
+  oneof response {
+    // The server must send back this message in response to a message with the
+    // "request_headers" field set.
+    HeadersResponse request_headers = 1;
+
+    // The server must send back this message in response to a message with the
+    // "response_headers" field set.
+    HeadersResponse response_headers = 2;
+
+    // The server must send back this message in response to a message with
+    // the "request_body" field set.
+    BodyResponse request_body = 3;
+
+    // The server must send back this message in response to a message with
+    // the "response_body" field set.
+    BodyResponse response_body = 4;
+
+    // The server must send back this message in response to a message with
+    // the "request_trailers" field set.
+    TrailersResponse request_trailers = 5;
+
+    // The server must send back this message in response to a message with
+    // the "response_trailers" field set.
+    TrailersResponse response_trailers = 6;
+
+    // If specified, return the specified response downstream to the client
+    // and end further filter processing on the request path.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Add new metadata to the Envoy context
+  google.protobuf.Struct new_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
```

Comment:
Per my other comment, I can buy that the server wants to change what is sent based on what it received, but do we need mode switching? This seems pretty complicated and it's unclear to me if it's really necessary?

Reply:
I'd be thrilled to get rid of the whole async vs sync mode thing for now, and also get rid of the idea of a permanent "mode change" driven by the server. We can bring in async mode in the future when we have made progress on the other stuff if it's turning out to be helpful.

Let me take a pass at simplifying this tomorrow and address the other comments too.